linode/manager
# feat: [M3-7128] - Coverage Report

## Description 📝
This PR (re)-introduces coverage reports for Cloud manager. It fixes the code coverage feature for Vitest and introduces a new action on PRs (non-draft only) that will do a coverage comparison between base branch (current state of develop) and the branch pointing to it. It will then comment on the PR (old VS new coverages) and the action will fail if coverage is lower. 

**NOTE**: this will not fail the CI if coverage is lower, just the action. It is meant to remain informational and not block the work however a lower coverage should be remedied whenever possible.

The GHA flows here follow recommended security conventions. 

- The first action, "Code Coverage", does not use any secret (since they are unavailable in forks PRs where this runs)
- The second action, "Coverage Comment" is based on completion of the first action and runs on the ref branch (develop). It does not uses any PAT since here we rely on the implicit GITHUB_TOKEN (see [docs](https://docs.github.com/en/actions/security-guides/automatic-token-authentication)) which is a one time auto-generated secret. 
- The third action (only running when master is merged) mimics what our `ci` job does to upload the storybook build to a bucket. In this case it just uploads an SVG badge to be displayed on our README. 

**IMPORTANT**
❌  this failure in the "Code Coverage" action **is expected**. Since the yarn command does not exist yet in develop, the action will start running properly when this PR is merged. To see a working version, refer to [my fork](https://github.com/abailly-akamai/manager/pull/14) where the code has been already merged to develop.

## Changes  🔄
- Fix the coverage generation
- Implement fine grained coverage exceptions
- implement three coverage GHA for reporting on open PRs and upload a badge.

## Preview 📷

### Commenting
![Screenshot 2023-11-28 at 11 06 40](https://github.com/linode/manager/assets/130582365/9e311c76-a09d-417b-b1b6-23cd879a5d10)


See 
- https://github.com/abailly-akamai/manager/pull/14 
- https://github.com/abailly-akamai/manager/actions/runs/7006885372
- https://github.com/abailly-akamai/manager/actions/runs/7007174993/job/19060652635

where this code has been merged to develop and all you can see all actions runs.

### Coverage Badge
![Screenshot 2023-11-28 at 11 06 32](https://github.com/linode/manager/assets/130582365/5b295c9d-5aed-4a26-8905-07814187f658)


## How to test 🧪
Since those actions are meant to run in the CI

- run `yarn` to update your dependencies
- pull code locally and run `yarn coverage` - confirm test w/ coverage is ran and coverage report opens afterwards 

## As an Author I have considered 🤔

*Check all that apply*
